### ISSUE TITLE
Logical error: Expected 46044 >= 88088

### ISSUE THREAD

**Issue by tavplubix** [ClickHouse/ClickHouse#49793]
https://s3.amazonaws.com/clickhouse-test-reports/0/9bf6175919406ca9cfec4a4ec4c3f6ad5f9d83c4/stress_test__asan_.html

```
[ 47003 ] {9ba1b7f9-0838-4709-a345-fa9030b96af9} <Fatal> : Logical error: \'Expected 46044 >= 88088 (Buffer path: 00170_test/src/pmiznbhxqqaqabvrhusvpdqcltxxe, hash key: c76c4ec90ae7e3a6379e5ecb073af99d, file_offset_of_buffer_end: 88088, read_until_position: 746040, internal buffer end: 88088, read_
 [ 32760 ] {9ba1b7f9-0838-4709-a345-fa9030b96af9} <Fatal> : Logical error: \'Expected 46044 >= 73708 (Buffer path: 00170_test/src/pmiznbhxqqaqabvrhusvpdqcltxxe, hash key: c76c4ec90ae7e3a6379e5ecb073af99d, file_offset_of_buffer_end: 73708, read_until_position: 746040, internal buffer end: 73708, read_
 [ 32964 ] {9ba1b7f9-0838-4709-a345-fa9030b96af9} <Fatal> : Logical error: \'Expected 46044 >= 69608 (Buffer path: 00170_test/src/pmiznbhxqqaqabvrhusvpdqcltxxe, hash key: c76c4ec90ae7e3a6379e5ecb073af99d, file_offset_of_buffer_end: 69608, read_until_position: 746040, internal buffer end: 69608, read_
 [ 32574 ] {9ba1b7f9-0838-4709-a345-fa9030b96af9} <Fatal> : Logical error: \'Expected 46044 >= 53539 (Buffer path: 00170_test/src/pmiznbhxqqaqabvrhusvpdqcltxxe, hash key: c76c4ec90ae7e3a6379e5ecb073af99d, file_offset_of_buffer_end: 53539, read_until_position: 746040, internal buffer end: 53539, read_
 [ 32627 ] {9ba1b7f9-0838-4709-a345-fa9030b96af9} <Fatal> : Logical error: \'Expected 46044 >= 77989 (Buffer path: 00170_test/src/pmiznbhxqqaqabvrhusvpdqcltxxe, hash key: c76c4ec90ae7e3a6379e5ecb073af99d, file_offset_of_buffer_end: 77989, read_until_position: 746040, internal buffer end: 77989, read_
 [ 3576 ] {9ba1b7f9-0838-4709-a345-fa9030b96af9} <Fatal> : Logical error: \'Expected 46044 >= 88300 (Buffer path: 00170_test/src/pmiznbhxqqaqabvrhusvpdqcltxxe, hash key: c76c4ec90ae7e3a6379e5ecb073af99d, file_offset_of_buffer_end: 88300, read_until_position: 746040, internal buffer end: 88300, read_t
 [ 33404 ] {9ba1b7f9-0838-4709-a345-fa9030b96af9} <Fatal> : Logical error: \'Expected 46044 >= 63380 (Buffer path: 00170_test/src/pmiznbhxqqaqabvrhusvpdqcltxxe, hash key: c76c4ec90ae7e3a6379e5ecb073af99d, file_offset_of_buffer_end: 63380, read_until_position: 746040, internal buffer end: 63380, read_
 [ 50411 ] {} <Fatal> BaseDaemon: ########################################
 [ 50412 ] {} <Fatal> BaseDaemon: ########################################
 [ 50413 ] {} <Fatal> BaseDaemon: ########################################
 [ 50411 ] {} <Fatal> BaseDaemon: (version 23.5.1.1 (official build), build id: 1755D3501B59E7ECA3DDF2A01B2E3653E8AF230F) (from thread 47003) (query_id: 9ba1b7f9-0838-4709-a345-fa9030b96af9) (query: SELECT GeneralInterests AS k1, GoalsReached AS k2, count() AS c FROM test.hits GROUP BY k1, k2 ORDER 
 [ 50413 ] {} <Fatal> BaseDaemon: (version 23.5.1.1 (official build), build id: 1755D3501B59E7ECA3DDF2A01B2E3653E8AF230F) (from thread 32964) (query_id: 9ba1b7f9-0838-4709-a345-fa9030b96af9) (query: SELECT GeneralInterests AS k1, GoalsReached AS k2, count() AS c FROM test.hits GROUP BY k1, k2 ORDER 
 [ 50414 ] {} <Fatal> BaseDaemon: ########################################
 [ 50412 ] {} <Fatal> BaseDaemon: (version 23.5.1.1 (official build), build id: 1755D3501B59E7ECA3DDF2A01B2E3653E8AF230F) (from thread 33404) (query_id: 9ba1b7f9-0838-4709-a345-fa9030b96af9) (query: SELECT GeneralInterests AS k1, GoalsReached AS k2, count() AS c FROM test.hits GROUP BY k1, k2 ORDER 
 [ 50415 ] {} <Fatal> BaseDaemon: ########################################
 [ 50411 ] {} <Fatal> BaseDaemon: 
 [ 50414 ] {} <Fatal> BaseDaemon: (version 23.5.1.1 (official build), build id: 1755D3501B59E7ECA3DDF2A01B2E3653E8AF230F) (from thread 32760) (query_id: 9ba1b7f9-0838-4709-a345-fa9030b96af9) (query: SELECT GeneralInterests AS k1, GoalsReached AS k2, count() AS c FROM test.hits GROUP BY k1, k2 ORDER 
 [ 50413 ] {} <Fatal> BaseDaemon: 
 [ 50415 ] {} <Fatal> BaseDaemon: (version 23.5.1.1 (official build), build id: 1755D3501B59E7ECA3DDF2A01B2E3653E8AF230F) (from thread 32627) (query_id: 9ba1b7f9-0838-4709-a345-fa9030b96af9) (query: SELECT GeneralInterests AS k1, GoalsReached AS k2, count() AS c FROM test.hits GROUP BY k1, k2 ORDER 
 [ 50412 ] {} <Fatal> BaseDaemon: 
 [ 50414 ] {} <Fatal> BaseDaemon: 
 [ 50417 ] {} <Fatal> BaseDaemon: ########################################
 [ 50413 ] {} <Fatal> BaseDaemon: Stack trace: 0x00007f2b5891d00b 0x00007f2b588fc859 0x0000562eaac4631b 0x0000562eaac47181 0x0000562eb271b9f1 0x0000562ebb997d19 0x0000562ebb998826 0x0000562ebb99822f 0x0000562ebb97fb10 0x0000562ebb97e52e 0x0000562ebb9c9511 0x0000562ebb9caa8e 0x0000562ebb13a11f 0x0000
 [ 50412 ] {} <Fatal> BaseDaemon: Stack trace: 0x00007f2b5891d00b 0x00007f2b588fc859 0x0000562eaac4631b 0x0000562eaac47181 0x0000562eb271b9f1 0x0000562ebb997d19 0x0000562ebb998826 0x0000562ebb99822f 0x0000562ebb97fb10 0x0000562ebb97e52e 0x0000562ebb9c9511 0x0000562ebb9ab4eb 0x0000562e99e5a07b 0x0000
 [ 50411 ] {} <Fatal> BaseDaemon: Stack trace: 0x00007f2b5891d00b 0x00007f2b588fc859 0x0000562eaac4631b 0x0000562eaac47181 0x0000562eb271b9f1 0x0000562ebb997d19 0x0000562ebb998826 0x0000562ebb99822f 0x0000562ebb97fb10 0x0000562ebb97e52e 0x0000562ebb9c9511 0x0000562ebb9caa8e 0x0000562ebb13a11f 0x0000
 [ 50415 ] {} <Fatal> BaseDaemon: 
 [ 50414 ] {} <Fatal> BaseDaemon: Stack trace: 0x00007f2b5891d00b 0x00007f2b588fc859 0x0000562eaac4631b 0x0000562eaac47181 0x0000562eb271b9f1 0x0000562ebb997d19 0x0000562ebb998826 0x0000562ebb99822f 0x0000562ebb97fb10 0x0000562ebb97e52e 0x0000562ebb9c9511 0x0000562ebb9caa8e 0x0000562ebb13a11f 0x0000
 [ 50418 ] {} <Fatal> BaseDaemon: ########################################
 [ 50417 ] {} <Fatal> BaseDaemon: (version 23.5.1.1 (official build), build id: 1755D3501B59E7ECA3DDF2A01B2E3653E8AF230F) (from thread 32574) (query_id: 9ba1b7f9-0838-4709-a345-fa9030b96af9) (query: SELECT GeneralInterests AS k1, GoalsReached AS k2, count() AS c FROM test.hits GROUP BY k1, k2 ORDER 
 [ 50414 ] {} <Fatal> BaseDaemon: 3. raise @ 0x00007f2b5891d00b in ?
 [ 50411 ] {} <Fatal> BaseDaemon: 3. raise @ 0x00007f2b5891d00b in ?
 [ 50413 ] {} <Fatal> BaseDaemon: 3. raise @ 0x00007f2b5891d00b in ?
 [ 50418 ] {} <Fatal> BaseDaemon: (version 23.5.1.1 (official build), build id: 1755D3501B59E7ECA3DDF2A01B2E3653E8AF230F) (from thread 3576) (query_id: 9ba1b7f9-0838-4709-a345-fa9030b96af9) (query: SELECT GeneralInterests AS k1, GoalsReached AS k2, count() AS c FROM test.hits GROUP BY k1, k2 ORDER B
 [ 50415 ] {} <Fatal> BaseDaemon: Stack trace: 0x00007f2b5891d00b 0x00007f2b588fc859 0x0000562eaac4631b 0x0000562eaac47181 0x0000562eb271b9f1 0x0000562ebb997d19 0x0000562ebb998826 0x0000562ebb99822f 0x0000562ebb97fb10 0x0000562ebb97e52e 0x0000562ebb9c9511 0x0000562ebb9caa8e 0x0000562ebb13a11f 0x0000
 [ 50412 ] {} <Fatal> BaseDaemon: 3. raise @ 0x00007f2b5891d00b in ?
 [ 50414 ] {} <Fatal> BaseDaemon: 4. abort @ 0x00007f2b588fc859 in ?
 [ 50411 ] {} <Fatal> BaseDaemon: 4. abort @ 0x00007f2b588fc859 in ?
 [ 50418 ] {} <Fatal> BaseDaemon: 
 [ 50417 ] {} <Fatal> BaseDaemon: 
 [ 50415 ] {} <Fatal> BaseDaemon: 3. raise @ 0x00007f2b5891d00b in ?
 [ 50413 ] {} <Fatal> BaseDaemon: 4. abort @ 0x00007f2b588fc859 in ?
 [ 50418 ] {} <Fatal> BaseDaemon: Stack trace: 0x00007f2b5891d00b 0x00007f2b588fc859 0x0000562eaac4631b 0x0000562eaac47181 0x0000562eb271b9f1 0x0000562ebb997d19 0x0000562ebb998826 0x0000562ebb99822f 0x0000562ebb97fb10 0x0000562ebb97e52e 0x0000562ebb9c9511 0x0000562ebb9caa8e 0x0000562ebb13a11f 0x0000
 [ 50412 ] {} <Fatal> BaseDaemon: 4. abort @ 0x00007f2b588fc859 in ?
 [ 50417 ] {} <Fatal> BaseDaemon: Stack trace: 0x00007f2b5891d00b 0x00007f2b588fc859 0x0000562eaac4631b 0x0000562eaac47181 0x0000562eb271b9f1 0x0000562ebb997d19 0x0000562ebb998826 0x0000562ebb99822f 0x0000562ebb97fb10 0x0000562ebb97e52e 0x0000562ebb9c9511 0x0000562ebb9caa8e 0x0000562ebb13a11f 0x0000
 [ 50418 ] {} <Fatal> BaseDaemon: 3. raise @ 0x00007f2b5891d00b in ?
 [ 50415 ] {} <Fatal> BaseDaemon: 4. abort @ 0x00007f2b588fc859 in ?
 [ 50417 ] {} <Fatal> BaseDaemon: 3. raise @ 0x00007f2b5891d00b in ?
 [ 50418 ] {} <Fatal> BaseDaemon: 4. abort @ 0x00007f2b588fc859 in ?
 [ 50417 ] {} <Fatal> BaseDaemon: 4. abort @ 0x00007f2b588fc859 in ?
 [ 50411 ] {} <Fatal> BaseDaemon: 5. ./build_docker/./src/Common/Exception.cpp:42: DB::abortOnFailedAssertion(String const&) @ 0x0000000022f4131b in /usr/bin/clickhouse
 [ 50413 ] {} <Fatal> BaseDaemon: 5. ./build_docker/./src/Common/Exception.cpp:42: DB::abortOnFailedAssertion(String const&) @ 0x0000000022f4131b in /usr/bin/clickhouse
 [ 50415 ] {} <Fatal> BaseDaemon: 5. ./build_docker/./src/Common/Exception.cpp:42: DB::abortOnFailedAssertion(String const&) @ 0x0000000022f4131b in /usr/bin/clickhouse
 [ 50417 ] {} <Fatal> BaseDaemon: 5. ./build_docker/./src/Common/Exception.cpp:42: DB::abortOnFailedAssertion(String const&) @ 0x0000000022f4131b in /usr/bin/clickhouse
 [ 50418 ] {} <Fatal> BaseDaemon: 5. ./build_docker/./src/Common/Exception.cpp:42: DB::abortOnFailedAssertion(String const&) @ 0x0000000022f4131b in /usr/bin/clickhouse
 [ 50412 ] {} <Fatal> BaseDaemon: 5. ./build_docker/./src/Common/Exception.cpp:42: DB::abortOnFailedAssertion(String const&) @ 0x0000000022f4131b in /usr/bin/clickhouse
 [ 50414 ] {} <Fatal> BaseDaemon: 5. ./build_docker/./src/Common/Exception.cpp:42: DB::abortOnFailedAssertion(String const&) @ 0x0000000022f4131b in /usr/bin/clickhouse
 [ 50411 ] {} <Fatal> BaseDaemon: 6.1. inlined from ./build_docker/./src/Common/Exception.cpp:65: DB::handle_error_code(String const&, int, bool, std::vector<void*, std::allocator<void*>> const&)
 [ 50411 ] {} <Fatal> BaseDaemon: 6. ./build_docker/./src/Common/Exception.cpp:93: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000022f42181 in /usr/bin/clickhouse
 [ 50415 ] {} <Fatal> BaseDaemon: 6.1. inlined from ./build_docker/./src/Common/Exception.cpp:65: DB::handle_error_code(String const&, int, bool, std::vector<void*, std::allocator<void*>> const&)
 [ 50415 ] {} <Fatal> BaseDaemon: 6. ./build_docker/./src/Common/Exception.cpp:93: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000022f42181 in /usr/bin/clickhouse
 [ 50413 ] {} <Fatal> BaseDaemon: 6.1. inlined from ./build_docker/./src/Common/Exception.cpp:65: DB::handle_error_code(String const&, int, bool, std::vector<void*, std::allocator<void*>> const&)
 [ 50413 ] {} <Fatal> BaseDaemon: 6. ./build_docker/./src/Common/Exception.cpp:93: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000022f42181 in /usr/bin/clickhouse
 [ 50414 ] {} <Fatal> BaseDaemon: 6.1. inlined from ./build_docker/./src/Common/Exception.cpp:65: DB::handle_error_code(String const&, int, bool, std::vector<void*, std::allocator<void*>> const&)
 [ 50414 ] {} <Fatal> BaseDaemon: 6. ./build_docker/./src/Common/Exception.cpp:93: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000022f42181 in /usr/bin/clickhouse
 [ 50418 ] {} <Fatal> BaseDaemon: 6.1. inlined from ./build_docker/./src/Common/Exception.cpp:65: DB::handle_error_code(String const&, int, bool, std::vector<void*, std::allocator<void*>> const&)
 [ 50418 ] {} <Fatal> BaseDaemon: 6. ./build_docker/./src/Common/Exception.cpp:93: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000022f42181 in /usr/bin/clickhouse
 [ 50417 ] {} <Fatal> BaseDaemon: 6.1. inlined from ./build_docker/./src/Common/Exception.cpp:65: DB::handle_error_code(String const&, int, bool, std::vector<void*, std::allocator<void*>> const&)
 [ 50417 ] {} <Fatal> BaseDaemon: 6. ./build_docker/./src/Common/Exception.cpp:93: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000022f42181 in /usr/bin/clickhouse
 [ 50412 ] {} <Fatal> BaseDaemon: 6.1. inlined from ./build_docker/./src/Common/Exception.cpp:65: DB::handle_error_code(String const&, int, bool, std::vector<void*, std::allocator<void*>> const&)
 [ 50412 ] {} <Fatal> BaseDaemon: 6. ./build_docker/./src/Common/Exception.cpp:93: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000022f42181 in /usr/bin/clickhouse
 [ 50418 ] {} <Fatal> BaseDaemon: 7.1. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/string:1499: String::__is_long[abi:v15000]() const
 [ 50413 ] {} <Fatal> BaseDaemon: 7.1. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/string:1499: String::__is_long[abi:v15000]() const
 [ 50415 ] {} <Fatal> BaseDaemon: 7.1. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/string:1499: String::__is_long[abi:v15000]() const
 [ 50413 ] {} <Fatal> BaseDaemon: 7.2. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/string:2333: ~basic_string
 [ 50418 ] {} <Fatal> BaseDaemon: 7.2. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/string:2333: ~basic_string
 [ 50415 ] {} <Fatal> BaseDaemon: 7.2. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/string:2333: ~basic_string
 [ 50413 ] {} <Fatal> BaseDaemon: 7.3. inlined from ./build_docker/./src/Common/Exception.h:42: ~MessageMasked
 [ 50418 ] {} <Fatal> BaseDaemon: 7.3. inlined from ./build_docker/./src/Common/Exception.h:42: ~MessageMasked
 [ 50415 ] {} <Fatal> BaseDaemon: 7.3. inlined from ./build_docker/./src/Common/Exception.h:42: ~MessageMasked
 [ 50413 ] {} <Fatal> BaseDaemon: 7.4. inlined from ./build_docker/./src/Common/Exception.h:54: Exception
 [ 50418 ] {} <Fatal> BaseDaemon: 7.4. inlined from ./build_docker/./src/Common/Exception.h:54: Exception
 [ 50417 ] {} <Fatal> BaseDaemon: 7.1. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/string:1499: String::__is_long[abi:v15000]() const
 [ 50412 ] {} <Fatal> BaseDaemon: 7.1. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/string:1499: String::__is_long[abi:v15000]() const
 [ 50415 ] {} <Fatal> BaseDaemon: 7.4. inlined from ./build_docker/./src/Common/Exception.h:54: Exception
 [ 50413 ] {} <Fatal> BaseDaemon: 7. ./build_docker/./src/Common/Exception.h:81: DB::Exception::Exception<unsigned long&, unsigned long&, String>(int, FormatStringHelperImpl<std::type_identity<unsigned long&>::type, std::type_identity<unsigned long&>::type, std::type_identity<String>::type>, unsigne
 [ 50418 ] {} <Fatal> BaseDaemon: 7. ./build_docker/./src/Common/Exception.h:81: DB::Exception::Exception<unsigned long&, unsigned long&, String>(int, FormatStringHelperImpl<std::type_identity<unsigned long&>::type, std::type_identity<unsigned long&>::type, std::type_identity<String>::type>, unsigne
 [ 50415 ] {} <Fatal> BaseDaemon: 7. ./build_docker/./src/Common/Exception.h:81: DB::Exception::Exception<unsigned long&, unsigned long&, String>(int, FormatStringHelperImpl<std::type_identity<unsigned long&>::type, std::type_identity<unsigned long&>::type, std::type_identity<String>::type>, unsigne
 [ 50417 ] {} <Fatal> BaseDaemon: 7.2. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/string:2333: ~basic_string
 [ 50414 ] {} <Fatal> BaseDaemon: 7.1. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/string:1499: String::__is_long[abi:v15000]() const
 [ 50412 ] {} <Fatal> BaseDaemon: 7.2. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/string:2333: ~basic_string
 [ 50417 ] {} <Fatal> BaseDaemon: 7.3. inlined from ./build_docker/./src/Common/Exception.h:42: ~MessageMasked
 [ 50414 ] {} <Fatal> BaseDaemon: 7.2. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/string:2333: ~basic_string
 [ 50412 ] {} <Fatal> BaseDaemon: 7.3. inlined from ./build_docker/./src/Common/Exception.h:42: ~MessageMasked
 [ 50417 ] {} <Fatal> BaseDaemon: 7.4. inlined from ./build_docker/./src/Common/Exception.h:54: Exception
 [ 50411 ] {} <Fatal> BaseDaemon: 7.1. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/string:1499: String::__is_long[abi:v15000]() const
 [ 50414 ] {} <Fatal> BaseDaemon: 7.3. inlined from ./build_docker/./src/Common/Exception.h:42: ~MessageMasked
 [ 50417 ] {} <Fatal> BaseDaemon: 7. ./build_docker/./src/Common/Exception.h:81: DB::Exception::Exception<unsigned long&, unsigned long&, String>(int, FormatStringHelperImpl<std::type_identity<unsigned long&>::type, std::type_identity<unsigned long&>::type, std::type_identity<String>::type>, unsigne
 [ 50412 ] {} <Fatal> BaseDaemon: 7.4. inlined from ./build_docker/./src/Common/Exception.h:54: Exception
 [ 50414 ] {} <Fatal> BaseDaemon: 7.4. inlined from ./build_docker/./src/Common/Exception.h:54: Exception
 [ 50411 ] {} <Fatal> BaseDaemon: 7.2. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/string:2333: ~basic_string
```

**Comment by tavplubix** [comment]
https://s3.amazonaws.com/clickhouse-test-reports/0/b6b9ccc76d0176c65a6c015ab2ffb4dc66579b4c/stress_test__ubsan_.html

**Comment by tavplubix** [comment]
https://s3.amazonaws.com/clickhouse-test-reports/0/32d422109783796a196ac9684311ec31be651b1d/stress_test__tsan_.html

**Comment by azat** [comment]
#50025 had been merged, could be closed?

**Comment by kssenii** [comment]
> https://github.com/ClickHouse/ClickHouse/pull/50025 had been merged, could be closed?

See comment https://github.com/ClickHouse/ClickHouse/issues/50468#issuecomment-1573473806.
In short: I believe error `Expected n > k` will NOT happen anymore just because of the assertion, which I added recently, - from https://github.com/ClickHouse/ClickHouse/issues/50468 - it happens for the same reason(bug) and happens earlier, not letting `Expected n > k` to happen. But the reason is not fixed yet, so I did not close this issue.
